File: One-to-One-Video/Wuji-macOS-Tutorial-Objective-C-1to1/Wuji-Mac-Tutorial-Objective-C/DeviceSelectionViewController.m
//
//  DeviceSelectionViewController.m
//  Wuji Mac Tutorial Objective-C
//
//  Created by 3 on 2020/12/17.
//  Portions Copyright (c) 2020 wuji-co. All rights reserved.
//

#import "DeviceSelectionViewController.h"

@interface DeviceSelectionViewController ()

@property (weak) IBOutlet NSPopUpButton *microphoneSelection;
@property (weak) IBOutlet NSPopUpButton *speakerSelection;
@property (weak) IBOutlet NSPopUpButton *cameraSelection;

@property (strong, nonatomic) NSArray* connectedRecordingDevices;
@property (strong, nonatomic) NSArray* connectedPlaybackDevices;
@property (strong, nonatomic) NSArray* connectedVideoCaptureDevices;

@end

@implementation DeviceSelectionViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.preferredContentSize = NSMakeSize(500, 250);
    [self loadDevicesInPopUpButtons];
}

- (void)loadDevicesInPopUpButtons {
    [self.microphoneSelection removeAllItems];
    [self.speakerSelection removeAllItems];
    [self.cameraSelection removeAllItems];

    self.connectedRecordingDevices = [self.wujiKit enumerateDevices:0];
    for (id device in self.connectedRecordingDevices) {
        [self.microphoneSelection addItemWithTitle:([device deviceName])];
    }
    
    self.connectedPlaybackDevices = [self.wujiKit enumerateDevices:1];
    for (id device in self.connectedPlaybackDevices) {
        [self.speakerSelection addItemWithTitle:([device deviceName])];
    }
    
    self.connectedVideoCaptureDevices = [self.wujiKit enumerateDevices:3];
    for (id device in self.connectedVideoCaptureDevices) {
        [self.cameraSelection addItemWithTitle:([device deviceName])];
    // Populate the NSPopUpButtons with the enumerated device list
    }
}

- (void)rtcEngine:(WujiRtcEngineKit *)engine device:(NSString*) deviceId type:(WujiMediaDeviceType) deviceType stateChanged:(NSInteger) state {
    [self loadDevicesInPopUpButtons];
    // Repopulate NSPopUpButtons if a device is plugged in or removed
}

- (IBAction)didClickConfirmButton:(NSButton *)button {
    [self.wujiKit setDevice:0 deviceId:[self.connectedRecordingDevices[self.microphoneSelection.indexOfSelectedItem] deviceId]];
    [self.wujiKit setDevice:1 deviceId:[self.connectedPlaybackDevices[self.speakerSelection.indexOfSelectedItem] deviceId]];
    [self.wujiKit setDevice:3 deviceId:[self.connectedVideoCaptureDevices[self.cameraSelection.indexOfSelectedItem] deviceId]];
    // Set the devices
    [self dismissViewController:self];
}

@end

File: One-to-One-Video/Wuji-macOS-Tutorial-Swift-1to1/Wuji-Mac-Tutorial-Swift/DeviceSelectionViewController.swift
//
//  DeviceSelectionViewController.swift
//  Wuji Mac Tutorial Swift
//
//  Created by 3 on 2020/12/18.
//  Copyright © 2020 Wuji.io. All rights reserved.
//

import Cocoa
import WujiRTCFramework

class DeviceSelectionViewController: NSViewController {
    @IBOutlet weak var microphoneSelection: NSPopUpButton!
    @IBOutlet weak var speakerSelection: NSPopUpButton!
    @IBOutlet weak var cameraSelection: NSPopUpButton!
    
    var connectedRecordingDevices: [WujiRtcDeviceInfo]!
    var connectedPlaybackDevices: [WujiRtcDeviceInfo]!
    var connectedVideoCaptureDevices: [WujiRtcDeviceInfo]!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.preferredContentSize = NSMakeSize(500, 250)
        loadDevicesInPopUpButtons()
    }
    
    func loadDevicesInPopUpButtons() {
        microphoneSelection.removeAllItems()
        speakerSelection.removeAllItems()
        cameraSelection.removeAllItems()
    
        connectedRecordingDevices = WujiKit.enumerateDevices(.audioRecording)
        for device in connectedRecordingDevices {
            if let title = device.deviceName {
                microphoneSelection.addItem(withTitle: title)
            }
        }
    
        connectedPlaybackDevices = WujiKit.enumerateDevices(.audioPlayout)
        for device in connectedPlaybackDevices {
            if let title = device.deviceName {
                speakerSelection.addItem(withTitle: title)
            }
        }
        
        connectedVideoCaptureDevices = WujiKit.enumerateDevices(.videoCapture)
        for device in connectedVideoCaptureDevices {
            if let title = device.deviceName {
                cameraSelection.addItem(withTitle: title)
            }
        }
        // Populate the NSPopUpButtons with the enumerated device list
    }
    
    @IBAction func didClickConfirmButton(_ sender: NSButton) {
        if let deviceId = connectedRecordingDevices[microphoneSelection.indexOfSelectedItem].deviceId {
            WujiKit.setDevice(.audioRecording, deviceId: deviceId)
        }
        if let deviceId = connectedPlaybackDevices[speakerSelection.indexOfSelectedItem].deviceId {
            WujiKit.setDevice(.audioPlayout, deviceId: deviceId)
        }
        if let deviceId = connectedVideoCaptureDevices[cameraSelection.indexOfSelectedItem].deviceId {
            WujiKit.setDevice(.videoCapture, deviceId: deviceId)
        }
        // Set the devices
        self.dismiss(self)
    }
}

extension DeviceSelectionViewController: WujiRtcEngineDelegate {
    @objc(rtcEngine:device:type:stateChanged:) func rtcEngine(_ engine: WujiRtcEngineKit, device deviceId: String, type deviceType: WujiMediaDeviceType, stateChanged state: Int) {
        loadDevicesInPopUpButtons()
        // Repopulate NSPopUpButtons if a device is plugged in or removed
    }
}
